compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp

```diff
@@ -0,0 +1,145 @@
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+
+#define DEBUG_TYPE "iree-amdaie-linalg-function-outlining"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Utility to outline the linalg compute op.
+static FailureOr<func::FuncOp> outlinedToAFunction(
+    IRRewriter &rewriter, ModuleOp moduleOp, linalg::LinalgOp computeOp,
+    std::string outlineFuncName) {
+  if (auto outlinedFuncOp = dyn_cast_if_present<func::FuncOp>(
+          moduleOp.lookupSymbol(outlineFuncName))) {
+    return outlinedFuncOp;
+  }
+
+  // Form outlined FunctionType.
+  SmallVector<Type> inputTypes = llvm::map_to_vector(
+      computeOp.getDpsInputs(), [](Value v) { return v.getType(); });
+  for (Value val : computeOp.getDpsInits()) inputTypes.push_back(val.getType());
+  auto outlinedFuncType =
+      FunctionType::get(rewriter.getContext(), inputTypes, /*outputTypes=*/{});
+
+  // Form outlined FuncSignature
+  rewriter.setInsertionPointToStart(moduleOp.getBody());
+  auto outlinedFunc = rewriter.create<func::FuncOp>(
+      moduleOp.getLoc(), outlineFuncName, outlinedFuncType);
+  outlinedFunc.setPrivate();
+
+  // Create an entry func block and map the original operands of the compute
+  // op to the block arguments.
+  Block *outlinedFuncBody = outlinedFunc.addEntryBlock();
+  rewriter.setInsertionPointToStart(outlinedFuncBody);
+  SmallVector<BlockArgument> outlinedFuncArgs = llvm::map_to_vector(
+      outlinedFunc.getArguments(), [&](BlockArgument bbArg) { return bbArg; });
+  unsigned bbArgIndex = 0;
+  IRMapping operandMap;
+  for (Value origOperand : computeOp.getDpsInputs())
+    operandMap.map(origOperand, outlinedFuncArgs[bbArgIndex++]);
+  for (Value origOperand : computeOp.getDpsInits())
+    operandMap.map(origOperand, outlinedFuncArgs[bbArgIndex++]);
+
+  // Clone the compute op while mapping the operand to the function block
+  // arguments.
+  Operation *clonedComputeOp = rewriter.clone(*computeOp, operandMap);
+
+  // Create terminator op returning the cloned compute op's results.
+  rewriter.setInsertionPointToEnd(outlinedFuncBody);
+  rewriter.create<func::ReturnOp>(clonedComputeOp->getLoc(), ValueRange({}));
+
+  return outlinedFunc;
+}
+
+class AMDAIELinalgFunctionOutliningPass
+    : public impl::AMDAIELinalgFunctionOutliningBase<
+          AMDAIELinalgFunctionOutliningPass> {
+ public:
+  AMDAIELinalgFunctionOutliningPass() = default;
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect, linalg::LinalgDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+void AMDAIELinalgFunctionOutliningPass::runOnOperation() {
+  ModuleOp moduleOp = getOperation();
+  MLIRContext *context = &getContext();
+  IRRewriter rewriter(context);
+
+  unsigned uniqueOutlinedMatmul = 0;
```

Comment:
Problem if this pass is run twice, as you'll create same names

Reply:
I don't think so, the `linalg.generic` for all Matmul won't even exist after the first run and the same applies for all Elementwise ops.

So, even if this pass is run more than once it won't cause any issue.